pkg/kkrtpsi/sender.go

```diff
@@ -0,0 +1,217 @@
+	stage1 := func() error {
+		logger.V(1).Info("Starting stage 1")
+
+		// sample Nhash hash seeds
+		for i := range seeds {
+			seeds[i] = make([]byte, hash.SaltLength)
+			if _, err := rand.Read(seeds[i]); err != nil {
+				return err
+			}
+			// write it into rw
+			if _, err := s.rw.Write(seeds[i]); err != nil {
+				return err
+			}
+		}
+
+		// read remote input size
+		if err := binary.Read(s.rw, binary.BigEndian, &remoteN); err != nil {
+			return err
+		}
+
+		// sample random 16 byte secret key for AES-128 and send to the receiver
+		sk := make([]byte, 16)
+		if _, err = rand.Read(sk); err != nil {
+			return err
+		}
+
+		// send the secret key
+		if _, err := s.rw.Write(sk); err != nil {
+			return err
+		}
+
+		// calculate number of OPRF from the receiver based on
+		// number of buckets in cuckooHashTable
+		oprfInputSize = int64(cuckoo.Factor * float64(remoteN))
+		if 1 > oprfInputSize {
+			oprfInputSize = 1
+		}
+
+		// exhaust local ids, and precompute all potential
+		// hashes and store them using the same
+		// cuckoo hash table parameters as the receiver.
+		go func() {
+			defer close(pseudorandIds)
+			cuckooHasher, err := cuckoo.NewCuckooHasher(uint64(remoteN), seeds)
+			if err != nil {
+				errChan <- err
+			}
+			// instantiate an AES block as well as a Highway Hash
+			aesBlock, err := aes.NewCipher(sk)
+			if err != nil {
+				errChan <- err
+			}
+
+			for id := range identifiers {
+				// hash and calculate pseudorandom code given each possible hash index
+				var bytes [cuckoo.Nhash][]byte
+				for hIdx := 0; hIdx < cuckoo.Nhash; hIdx++ {
+					bytes[hIdx] = crypto.PseudorandomCode(aesBlock, id, byte(hIdx))
+				}
+				pseudorandIds <- oprfEncodeInputs{prcEncoded: bytes, bucketIdx: cuckooHasher.BucketIndices(id)}
+			}
+			hasher := cuckooHasher.GetHasher()
+			hashChan <- hasher
+		}()
+
+		// end stage1
+		timer, mem = printStageStats(logger, 1, start, start, 0)
+		logger.V(1).Info("Finished stage 1")
+		return nil
+	}
+
+	// stage 2: act as sender in OPRF, and receive OPRF keys
+	stage2 := func() error {
+		logger.V(1).Info("Starting stage 2")
+
+		// instantiate OPRF sender with agreed parameters
+		oSender, err := oprf.NewOPRF(int(oprfInputSize))
+		if err != nil {
+			return err
+		}
+
+		oprfKeys, err = oSender.Send(s.rw)
+		if err != nil {
+			return err
+		}
+
+		// end stage2
+		timer, mem = printStageStats(logger, 2, timer, start, mem)
+		logger.V(1).Info("Finished stage 2")
+		return nil
+	}
+
+	// stage 3: compute all possible OPRF output using keys obtained from stage2
+	stage3 := func() error {
+		logger.V(1).Info("Starting stage 3")
+
+		// inform the receiver the number of local ID
+		if err := binary.Write(s.rw, binary.BigEndian, &n); err != nil {
+			return err
+		}
+
+		var hasher hash.Hasher
+		// read error
+		select {
+		case err := <-errChan:
+			if err != nil {
+				return err
+			}
+		// block until we have the hasher
+		case hasher = <-hashChan:
```

Comment:
Since we block until we get the hasher, this acts as a barrier and we don't need to use a channel of size `n` for the pseudorandomids. I think the simplest way of implementing that would be to define a function that returns a channel of `struct precomputeResult` which contains the hasher and the slice of pseudorandIds.

Then since we wait for this entire step to be completed here, we can pass the hasher and the entire slice to EncodeAndHashAllParallel and really simplify it. I'll comment in that function.